File: Maximum_Subarray_II.cpp
class Solution {
public:
    /**
     * @param nums: A list of integers
     * @return: An integer denotes the sum of max two non-overlapping subarrays
     */
    int maxTwoSubArrays(vector<int> nums) {
        // write your code here
        int len = nums.size();
        vector<int> max_LR(len), max_RL(len);
        int lr_max = INT_MIN, lr_sum = 0;
        for (int i = 0; i < len; ++i)
        {
            lr_sum += nums[i];
            lr_max = max(lr_max, lr_sum);
            lr_sum = max(0, lr_sum);
            max_LR[i] = lr_max;
        }
        
        int rl_max = INT_MIN, rl_sum = 0;
        for (int i = len - 1; i > -1; --i)
        {
            rl_sum += nums[i];
            rl_max = max(rl_max, rl_sum);
            rl_sum = max(0, rl_sum);
            max_RL[i] = rl_max;
        }
        
        int result = INT_MIN;
        for (int i = 0; i < len - 1; ++i)
        {
            result = max(result, max_LR[i] + max_RL[i + 1]);
        }
        return result;
    }
};

File: Maximum Subarray II.py
class Solution:
    """
    @param nums: A list of integers
    @return: An integer denotes the sum of max two non-overlapping subarrays
    """
    def maxTwoSubArrays(self, nums):
        # write your code here
        if not nums:
            return 0
        length = len(nums)
        leftMax = [0] * length
        local = float('-inf')
        gl = float('-inf')
        for i in range(length):
            local = max(local + nums[i], nums[i])
            gl = max(local, gl)
            leftMax[i] = gl
        
        rightMax = [0] * length
        local = float('-inf')
        gl = float('-inf')
        for i in range(length - 1, -1, -1):
            local = max(local + nums[i], nums[i])
            gl = max(gl, local)
            rightMax[i] = gl
        
        ans = float('-inf')
        for i in range(length - 1):
            ans = max(ans, leftMax[i] + rightMax[i + 1])
        return ans
